fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/PrefetchableBufferedChecksumIndexInput.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * PrefetchableBufferedChecksumIndexInput.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import com.apple.foundationdb.record.lucene.directory.FDBIndexInput;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+
+import java.io.IOException;
+
+/**
+ * An Index Input that attempts to keep a 160K buffer in front of the read at a minimum.  The goal for this
+ * class is to provide some parallelism for fetches from FDB when we are attempting to merge segments.  The signaling
+ * of when to prefetch occurs in the FDBIndexOutput#copyBytes method.
+ *
+ */
+public class PrefetchableBufferedChecksumIndexInput extends BufferedChecksumIndexInput {
+    private IndexInput main;
+    private static final long PREFETCH_MINIMUM_BUFFER = 16L * 1024L * 10L; // Attempt to keep a large enough buffer
+    /**
+     * Total number of bytes to prefetch, usually set via FDBIndexOutput#copyBytes.
+     *
+     */
+    private long numberOfBytesToPrefetch;
+    /**
+     * Current active prefetch size.
+     *
+     */
+    private long currentPrefetch = 0;
+    /**
+     * Total number of bytes prefetched.
+     *
+     */
+    private long totalPrefetch = 0;
+
+    public PrefetchableBufferedChecksumIndexInput(final IndexInput main) {
+        super(main);
+        this.main = main;
+    }
+
+    public boolean supportsPrefetch() {
+        return main instanceof FDBIndexInput;
+    }
+
+    public void setNumberOfBytesToPrefetch(long numberOfBytesToPrefetch) {
+        this.numberOfBytesToPrefetch = numberOfBytesToPrefetch;
+        this.currentPrefetch = 0;
+        this.totalPrefetch = 0;
+        checkBuffer();
+    }
+
+    public void checkBuffer() {
+        if (numberOfBytesToPrefetch > 0) {
+            if (currentPrefetch < PREFETCH_MINIMUM_BUFFER && currentPrefetch < numberOfBytesToPrefetch) {
+                long bytesToFetch = Math.min(PREFETCH_MINIMUM_BUFFER - currentPrefetch, numberOfBytesToPrefetch);
+                prefetch(bytesToFetch, totalPrefetch);
+                currentPrefetch += bytesToFetch;
+                totalPrefetch += bytesToFetch;
+            }
+        }
+
+    }
+
+    public void prefetch(long numberOfBytes, long currentPrefetch) {
```

Comment:
Could you rename the argument ```currentPrefetch```? looks like it is passed from the global parameter ```totalPrefetch```, but also conflicts with the global parameter ```currentPrefetch```, which might be a little confusing.

Reply:
Removed this...